modules/git/foreachref/parser.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package foreachref
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"strings"
+)
+
+// Parser parses 'git for-each-ref' output according to a given output Format.
+type Parser struct {
+	//  tokenizes 'git for-each-ref' output into "reference paragraphs".
+	scanner *bufio.Scanner
+
+	// format represents the '--format' string that describes the expected
+	// 'git for-each-ref' output structure.
+	format Format
+
+	// err holds the last encountered error during parsing.
+	err error
+}
+
+// NewParser creates a 'git for-each-ref' output parser that will parse all
+// references in the provided Reader. The references in the output are assumed
+// to follow the specified Format.
+func NewParser(r io.Reader, format Format) *Parser {
+	scanner := bufio.NewScanner(r)
+
+	// Split input into delimiter-separated "reference blocks".
+	scanner.Split(
+		func(data []byte, atEOF bool) (advance int, token []byte, err error) {
+			// Scan until delimiter, marking end of reference.
+			delimIdx := bytes.Index(data, format.refDelim)
+			if delimIdx >= 0 {
+				token := data[:delimIdx]
+				advance := delimIdx + len(format.refDelim)
+				return advance, token, nil
+			}
+			// If we're at EOF, we have a final, non-terminated reference. Return it.
+			if atEOF {
+				return len(data), bytes.TrimSpace(data), nil
```

Comment:
```suggestion
				if trimmed := bytes.TrimSpace(data); len(trimmed) == 0 {
					// at the moment it seems that there are "\n" before EOF, in such case we do not need to process it 
					// this is a temporary solution to make the parser work correctly.
					return len(data), trimmed, nil
				}
				return len(data), data, nil
```

(if no conclusion, some comments in my mind)

Reply:
I did take another look and as it turns out `git for-each-ref` _always_ adds a newline after every reference it outputs. So I modified the scanner split function to take that into account and could skip the `TrimSpace` call thanks to that. Also, I updated the test cases to make sure newlines are added to every "simulated" `git for-each-ref` output. c1da4e0ae

Reply:
Great! So clear solution.